tls-client/HelloHttpsClient.cpp

```diff
@@ -0,0 +1,431 @@
+    mbedtls_ssl_conf_verify( &ssl_conf, sslVerify, NULL );
+    mbedtls_ssl_conf_dbg( &ssl_conf, sslDebug, NULL );
+    mbedtls_debug_set_threshold( HELLO_HTTPS_CLIENT_DEBUG_LEVEL );
+#endif /* HELLO_HTTPS_CLIENT_DEBUG_LEVEL > 0 */
+
+    ret = mbedtls_ssl_setup( &ssl, &ssl_conf );
+    if( ret != 0 )
+    {
+        logTlsError( "mbedtls_ssl_setup()", ret );
+        return( ret );
+    }
+
+    ret = mbedtls_ssl_set_hostname( &ssl, server_name );
+    if( ret != 0 )
+    {
+        logTlsError( "mbedtls_ssl_set_hostname()", ret );
+        return( ret );
+    }
+
+    mbedtls_ssl_set_bio( &ssl, static_cast<void *>( socket ), sslSend,
+                         sslRecv, NULL );
+
+    return( 0 );
+}
+
+int HelloHttpsClient::sslRecv( void *ctx, unsigned char *buf, size_t len )
+{
+    TCPSocket *socket = static_cast<TCPSocket *>( ctx );
+    int ret = socket->recv( buf, len );
+
+    if( ret == NSAPI_ERROR_WOULD_BLOCK )
+    {
+        return( MBEDTLS_ERR_SSL_WANT_READ );
+    }
+    else if( ret < 0 )
+    {
+        mbedtls_printf( "Failed call to socket->recv(): %d\r\n", ret );
+    }
+
+    return( ret );
+}
+
+int HelloHttpsClient::sslSend( void *ctx, const unsigned char *buf, size_t len )
+{
+    TCPSocket *socket = static_cast<TCPSocket *>( ctx );
+    int ret = socket->send( buf, len );
+
+    if( ret == NSAPI_ERROR_WOULD_BLOCK )
+    {
+        return( MBEDTLS_ERR_SSL_WANT_WRITE );
+    }
+    else if( ret < 0 )
+    {
+        mbedtls_printf( "Failed call to socket->send(): %d\r\n", ret );
+    }
+
+    return( ret );
+}
+
+void HelloHttpsClient::sslDebug( void *ctx, int level, const char *file,
+                                 int line, const char *str )
+{
+    (void)ctx;
+
+    const char *p, *basename;
+
+    /* Extract basename from file */
+    for( p = basename = file; *p != '\0'; p++ )
+    {
+        if( *p == '/' || *p == '\\' )
+        {
+            basename = p + 1;
+        }
+    }
+
+    mbedtls_printf( "%s:%d: |%d| %s\r", basename, line, level, str );
+}
+
+int HelloHttpsClient::sslVerify( void *data, mbedtls_x509_crt *crt, int depth,
+                                 uint32_t *flags )
+{
+    (void)data;
+
+    int ret = -1;
+    char *buf = new char[GENERAL_PURPOSE_BUFFER_LENGTH];
+
+    if( buf == NULL )
+    {
+        mbedtls_printf( "Failed to allocate sslVerify() buffer\r\n" );
+        goto exit;
+    }
+
+    ret = mbedtls_x509_crt_info( buf, GENERAL_PURPOSE_BUFFER_LENGTH,
+                                 "\r  ", crt );
+    if( ret < 0 )
+    {
+        mbedtls_printf( "Failed call to mbedtls_x509_crt_info(): -0x%04X\r\n",
+                        ret );
+        goto cleanup;
+    }
+    else
+    {
+        mbedtls_printf( "Verifying certificate at depth %d:\r\n%s\r\n",
+                        depth, buf );
+    }
+
+    ret = 0;
+
+cleanup:
+    delete[] buf;
+
+exit:
+    return( ret );
+}
+
+int HelloHttpsClient::run( void )
+{
+    int ret;
+    size_t req_len, req_offset, resp_offset;
```

Comment:
C++ does not enforce declaring variables at top. They should be as close to the usage as possible. These could declared and assigned at https://github.com/ARMmbed/mbed-os-example-tls/pull/63/files#diff-39d7102cc3f72bf4a2240cc439f38d0fR388
This also avoids uninitialised use.